File: panel/app/services/mail.py
"""邮件发送服务。

SMTP 失败进入内存队列重试 3 次；开发环境未配置 SMTP 时，
将邮件内容打印到日志以便联调。
"""
from __future__ import annotations

import asyncio
import smtplib
import time
from email.mime.text import MIMEText
from email.utils import formataddr

from ..core.config import get_settings
from ..core.logger import get_logger

logger = get_logger("weavenet.mail")

settings = get_settings()

MAX_RETRIES = 3


class MailQueue:
    """内存邮件队列 + 异步发送。"""

    def __init__(self) -> None:
        self._queue: asyncio.Queue[dict] = asyncio.Queue(maxsize=1000)
        self._worker: asyncio.Task | None = None

    def start(self) -> None:
        if self._worker is None or self._worker.done():
            self._worker = asyncio.create_task(self._run())

    async def stop(self) -> None:
        if self._worker:
            self._worker.cancel()
            try:
                await self._worker
            except asyncio.CancelledError:
                pass
            self._worker = None

    async def _run(self) -> None:
        while True:
            item = await self._queue.get()
            to_email, subject, body = (
                item["to"],
                item["subject"],
                item["body"],
            )
            ok = await asyncio.to_thread(self._send_once, to_email, subject, body)
            retries = item.get("retries", 0)
            if not ok and retries < MAX_RETRIES:
                item["retries"] = retries + 1
                logger.warning("邮件发送失败，入队重试 %s/%s: %s", retries + 1, MAX_RETRIES, to_email)
                await asyncio.sleep(2)
                await self._queue.put(item)
            elif not ok:
                logger.error("邮件发送最终失败: %s", to_email)
            self._queue.task_done()

    def _send_once(self, to_email: str, subject: str, body: str) -> bool:
        if not settings.smtp_host or not settings.smtp_from:
            # 开发环境未配置 SMTP，打印到日志
            logger.info("[邮件占位] 收件人=%s 主题=%s 内容=%s", to_email, subject, body[:200])
            return True
        try:
            msg = MIMEText(body, "plain", "utf-8")
            msg["Subject"] = subject
            msg["From"] = formataddr((settings.app_name, settings.smtp_from))
            msg["To"] = to_email
            if settings.smtp_use_ssl:
                server = smtplib.SMTP_SSL(
                    settings.smtp_host, settings.smtp_port, timeout=15
                )
            else:
                server = smtplib.SMTP(
                    settings.smtp_host, settings.smtp_port, timeout=15
                )
                if settings.smtp_use_tls:
                    server.starttls()
            if settings.smtp_user:
                server.login(settings.smtp_user, settings.smtp_password)
            server.sendmail(settings.smtp_from, [to_email], msg.as_string())
            server.quit()
            return True
        except Exception as exc:
            logger.warning("SMTP 发送异常: %s", exc)
            return False

    def send(self, to_email: str, subject: str, body: str) -> None:
        """入队发送（非阻塞）。"""
        self.start()
        try:
            self._queue.put_nowait({"to": to_email, "subject": subject, "body": body, "retries": 0})
        except asyncio.QueueFull:
            logger.error("邮件队列已满，丢弃邮件: %s", to_email)


mail_queue = MailQueue()


def send_verification_code(to_email: str, code: str, purpose: str) -> None:
    """发送验证码邮件。"""
    purpose_text = {
        "register": "账号注册激活",
        "reset_password": "找回密码",
        "change_email": "修改邮箱",
    }.get(purpose, "账号验证")
    subject = f"【{settings.app_name}】{purpose_text}验证码"
    body = (
        f"您好：\n\n"
        f"您在 {settings.app_name} 的{purpose_text}验证码为：{code}\n"
        f"验证码自发送起 5 分钟内有效，请及时完成验证。\n\n"
        f"如非本人操作，请忽略本邮件。\n"
        f"—— {settings.app_name} 团队"
    )
    mail_queue.send(to_email, subject, body)

File: panel-go/services/mail.go
package services

import (
	"crypto/tls"
	"fmt"
	"log"
	"net/smtp"
	"strings"
	"time"

	"weavenet/panel/config"
)

// mailItem 待发送邮件。
type mailItem struct {
	to      string
	subject string
	body    string
}

// mailQueue 内存邮件队列。
var mailQueue = make(chan mailItem, 1000)
var mailStarted = false

// StartMailWorker 启动邮件消费 worker。
func StartMailWorker() {
	if mailStarted {
		return
	}
	mailStarted = true
	go func() {
		for item := range mailQueue {
			for attempt := 0; attempt < 3; attempt++ {
				if err := sendMail(item); err == nil {
					break
				}
				time.Sleep(2 * time.Second)
			}
		}
	}()
	log.Println("[mail] 邮件 worker 已启动")
}

// StopMailWorker 停止邮件 worker（关闭队列）。
func StopMailWorker() {
	if mailStarted {
		mailStarted = false
		close(mailQueue)
	}
}

// smtpConfigured 是否配置了 SMTP。
func smtpConfigured() bool {
	return config.C.SmtpHost != "" && config.C.SmtpFrom != ""
}

// sendMail 发送一封邮件（同步）。
func sendMail(item mailItem) error {
	if !smtpConfigured() {
		log.Printf("[邮件占位] to=%s subject=%s body=%s", item.to, item.subject, item.body)
		return nil
	}
	host := config.C.SmtpHost
	port := config.C.SmtpPort
	addr := fmt.Sprintf("%s:%d", host, port)
	from := config.C.SmtpFrom
	user := config.C.SmtpUser
	pass := config.C.SmtpPassword

	msg := "From: " + config.C.AppName + " <" + from + ">\r\n" +
		"To: " + item.to + "\r\n" +
		"Subject: " + item.subject + "\r\n" +
		"MIME-Version: 1.0\r\n" +
		"Content-Type: text/plain; charset=UTF-8\r\n\r\n" +
		item.body

	var conn *smtp.Client
	var err error
	if config.C.SmtpUseSSL {
		tlsCfg := &tls.Config{ServerName: host, InsecureSkipVerify: false}
		tconn, terr := tls.Dial("tcp", addr, tlsCfg)
		if terr != nil {
			return terr
		}
		conn, err = smtp.NewClient(tconn, host)
	} else {
		conn, err = smtp.Dial(addr)
	}
	if err != nil {
		return err
	}
	defer conn.Close()
	if config.C.SmtpUseTLS && !config.C.SmtpUseSSL {
		if err := conn.StartTLS(&tls.Config{ServerName: host}); err != nil {
			return err
		}
	}
	if user != "" {
		if err := conn.Auth(smtp.PlainAuth("", user, pass, host)); err != nil {
			return err
		}
	}
	if err := conn.Mail(from); err != nil {
		return err
	}
	if err := conn.Rcpt(item.to); err != nil {
		return err
	}
	w, err := conn.Data()
	if err != nil {
		return err
	}
	if _, err := w.Write([]byte(msg)); err != nil {
		return err
	}
	return w.Close()
}

// EnqueueMail 入队邮件（队列满则丢弃）。
func EnqueueMail(to, subject, body string) {
	select {
	case mailQueue <- mailItem{to: to, subject: subject, body: body}:
	default:
		log.Printf("[mail] 邮件队列已满，丢弃: %s", subject)
	}
}

// SendVerificationCode 发送验证码邮件。
func SendVerificationCode(to, code, purpose string) {
	purposeText := map[string]string{
		"register":       "账号注册激活",
		"reset_password": "找回密码",
		"change_email":   "修改邮箱",
	}[purpose]
	if purposeText == "" {
		purposeText = purpose
	}
	subject := fmt.Sprintf("【%s】%s验证码", config.C.AppName, purposeText)
	body := fmt.Sprintf("您好，您正在%s。\n\n验证码：%s\n\n验证码 5 分钟内有效，请勿泄露给他人。\n\n如非本人操作，请忽略此邮件。", purposeText, code)
	EnqueueMail(to, subject, body)
}

// FormatSMTPConfig 将配置写入 system_configs（占位，供 admin 读取）。
func FormatSMTPConfig() map[string]any {
	return map[string]any{
		"smtp_host":     config.C.SmtpHost,
		"smtp_port":     config.C.SmtpPort,
		"smtp_user":     config.C.SmtpUser,
		"smtp_from":     config.C.SmtpFrom,
		"smtp_use_ssl":  config.C.SmtpUseSSL,
		"smtp_use_tls":  config.C.SmtpUseTLS,
		"smtp_password": strings.Repeat("*", len(config.C.SmtpPassword)),
	}
}
